test/e2e/background-socket/socket-background-to-mocha.ts

```diff
@@ -0,0 +1,136 @@
+import log from 'loglevel';
+import { WindowProperties } from './window-handles';
+
+/**
+ * This singleton class runs on the Extension background script (service worker in MV3).
+ * It's used to communicate from the Extension background script to the Mocha/Selenium test.
+ * The main advantage is that it can call chrome.tabs.query().
+ * We had hoped it would be able to call chrome.tabs.highlight(), but Selenium doesn't see the tab change.
+ */
+class SocketBackgroundToMocha {
+  private client: WebSocket;
+
+  constructor() {
+    this.client = new WebSocket('ws://localhost:8111');
+
+    this.client.onopen = () => log.debug('WebSocket connection opened');
+
+    this.client.onmessage = (event: MessageEvent) =>
+      this.receivedMessage(event.data);
+
+    this.client.onclose = () => log.debug('WebSocket connection closed');
+
+    this.client.onerror = (error) => log.error('WebSocket error:', error);
+  }
+
+  /**
+   * Waits until a window with the given property is open.
+   * delayStep = 200ms, timeout = 10s
+   *
+   * You can think of this kind of like a template function:
+   * If `property` is `title`, then this becomes `waitUntilWindowWithTitle`
+   * If `property` is `url`, then this becomes `waitUntilWindowWithUrl`
+   * Remember that `a[property]` becomes `a.title` or `a.url`
+   *
+   * @param property - 'title' or 'url'
+   * @param value - The value we're searching for and want to wait for
+   * @returns The handle of the window tab with the given property value
+   */
+  async waitUntilWindowWithProperty(property: WindowProperties, value: string) {
+    let tabs: chrome.tabs.Tab[] = [];
+    const delayStep = 200;
+    const timeout = 10000;
+
+    for (
+      let timeElapsed = 0;
+      timeElapsed <= timeout;
+      timeElapsed += delayStep
+    ) {
+      tabs = await this.queryTabs({});
+
+      const index = tabs.findIndex((a) => a[property] === value);
+
+      if (index !== -1) {
+        this.send({ command: 'openTabs', tabs: this.cleanTabs(tabs) });
+        return;
+      }
+
+      // wait for delayStep milliseconds
+      await new Promise((resolve) => setTimeout(resolve, delayStep));
+    }
+
+    // The window was not found at the end of the timeout
+    this.send({ command: 'notFound', tabs: this.cleanTabs(tabs) });
+  }
+
+  // This function exists to support both MV2 and MV3
+  private async queryTabs(queryInfo: object): Promise<chrome.tabs.Tab[]> {
+    if (
+      process.env.ENABLE_MV3 === 'true' ||
+      process.env.ENABLE_MV3 === undefined
+    ) {
+      // With MV3, chrome.tabs.query has an await form
+      return await chrome.tabs.query(queryInfo);
+    }
+
+    // With MV2, we have to wrap chrome.tabs.query in a Promise
+    return new Promise((resolve) => {
+      chrome.tabs.query(queryInfo, (tabs: chrome.tabs.Tab[]) => {
+        resolve(tabs);
+      });
+    });
+  }
```

Comment:
Did you explore using the `chrome.tabs.on*` events to automatically send the browser's tab state to the test runner? This seems like it could let us get rid of polling entirely. The `onUpdated` event fires whenever a tab's title changes, for example.

Reply:
We discussed this in Slack.  Synchronizing and cross-referencing the tabs lists between the Extension and Selenium is likely to be very challenging, because the handles are different and the order is different.  I'm going to attempt this in a follow-up though.